feat: Focus Management within ShadowDOM

MahmoudElsayad opened the pull request:
Closes #1472

This PR enhances focus management capabilities in React Spectrum applications when used within Shadow DOM environments.

## Changes
- Introduced a new utility function `getRootNode`, designed to return a given element's contextually appropriate root (Document or ShadowRoot). This improves the library's ability to query and manipulate focus within shadow DOMs.
- Added Shadow DOM support to the following:
  - `FocusScope`
  - `useFocus`
  - `useFocusVisible`
  - `useFocusWithin`
  - `useInteractionOutside`
  - `usePress`
- Implemented a new utility function, `getRootBody` that determines the effective "body" element for an event's propagation path, supporting both Shadow DOM and traditional document structures.
- implemented a new utility, ' getDeepActiveElement,` which retrieves the currently focused element across Shadow DOM boundaries, ensuring accurate focus management in complex DOM structures.
- Added tests for all new utility functions and affected hooks/components.

## ✅ Pull Request Checklist:

- [x] Included link to corresponding [React Spectrum GitHub Issue](https://github.com/adobe/react-spectrum/issues).
- [x] Added/updated unit tests and storybook for this change (for new code or code which already has tests).
- [x] Filled out test instructions.
- [ ] Updated documentation (if it already exists for this component).
- [ ] Looked at the Accessibility Practices for this feature - [Aria Practices](https://www.w3.org/WAI/ARIA/apg/)

## 📝 Test Instructions:
- Open any storybook example that uses any of the affected hooks/components
- Test the functionality and make sure it works when the story is encapsulated inside a shadow root.

```js
// custom wrapper to run any story inside a shadow root
function ShadowDomWrapper({ children }) {
  const container = useRef(null);

  useEffect(() => {
    const _container = container.current;
    if (!_container) return;

    const div = document.createElement('div');
    _container.appendChild(div);

    const shadowRoot = div.attachShadow({ mode: 'open' });
    const main = document.createElement('main');
    shadowRoot.appendChild(main);

    // Wrap children with the ThemeProvider when rendering
    ReactDOM.render(
      children,
      main
    );

    return () => {
      ReactDOM.unmountComponentAtNode(main);
      if (_container) {
        _container.innerHTML = '';
      }
    };
  }, [children]);

  return <div ref={container} />;
}
```

```diff
// .storybook/preview.js
export const decorators = [
+    (Story) => (
+     <ShadowDomWrapper>
+        <Story />
+      </ShadowDomWrapper>
+    ),
  withScrollingSwitcher,
  ...(process.env.NODE_ENV !== 'production' ? [withStrictModeSwitcher] : []),
  withProviderSwitcher
];
```
<!--- Include instructions to test this pull request -->

## 🧢 Your Project:
PSPDFKit - <img src="https://pspdfkit.com/assets/images/logos/pspdfkit-3ccd8253.svg"  width="20"> 

<!--- Company/project for pull request -->

ritz078 commented:
@snowystinger We are working on fixing the linting and type errors but if you want you can give an early eye to this PR.

yihuiliao commented:
You'll need to sign the [Adobe CLA ](https://opensource.adobe.com/cla.html) and then close and reopen this PR for the it to pass. 

But also I'd like to let you know that we might be slow to review as we work through some other priorities and we apologize for the wait.

davidferguson commented:
This is a great contribution, and it would be very useful for our team to see Shadow DOM support in React Spectrum.

I've had a play with it, and can see it works for some components, however it seems to not work with all overlays (AlertDialog, ContextualHelp, ActionMenu, Picker, etc.). There are a couple of issues, including that after closing the overlay, focus is not returned correctly to the element that triggered the open, and there also seems to be some state bug too where it thinks the overlay is still open.

See this screen recording for an example with Picker - you will see that focus is not returned after choosing a value, and the Picker trigger still looks "pressed":

![ezgif-2-6a42c41408](https://github.com/adobe/react-spectrum/assets/13520342/27f38e85-7af9-4905-8df9-bd94a526e675)

I tried modifying `useRestoreFocus` of `FocusScope.tsx` to set `const nodeToRestoreRef` to be `getDeepActiveElement()`, and this fixed the issue of focus not being returned, however doesn't solve the state issue. In the following screen capture, you'll see that the focus ring is returned as it should be, however when I tab off the Picker, it still looks "pressed", and clicking on it to open it has no effect. I need to focus onto something else and hit enter to restore the state correctly:

![ezgif-2-4480c9e82b](https://github.com/adobe/react-spectrum/assets/13520342/c163f70e-a5be-4930-b4f8-f416ae4b81b8)

I've not had too much of a dig around, but I can see there are quite a lot of uses of `.activeElement` in React Spectrum (I count 63 uses across 27 files, 51 of which are `document.activeElement`). I wonder if some/all of these also need to be updated to get the actual active element, rather than assuming it's the document's active element?

MahmoudElsayad commented:
Thanks, @davidferguson, for looking at the PR; what you mentioned is really helpful, and finding the correct `activeElement` will, to some extent, solve most of the issues.

I will try to invest some time soon to fix the overlays raised issues and the state issue; most likely the overlay issues will share the same root cause.

davidferguson commented:
Hi @MahmoudElsayad, I did spend some time looking into this, and fixing it for the use cases that we needed.

I'm attaching a patch file in the hope it may be useful for you. This is based off 3.34.1, so not the current main, or your PR, but it should be similar to apply. It makes overlays (menus, dialogs) and toasts and landmarks functional for our use case.

[overlay-toast-fixes.patch.txt](https://github.com/adobe/react-spectrum/files/15204208/overlay-toast-fixes.patch.txt)

@yihuiliao and @snowystinger, is there a preferred approach of how you'd like to see webcomponent & React Spectrum support progress? Is working on small individual fixes like @MahmoudElsayad's work, and now my overlay work the way forward, or are you hoping for a more holistic, comprehensive approach that updates the entire library at once?

MahmoudElsayad commented:
@davidferguson Thank you for the patch! I think the way to go would be in small individual fixes; it will be great if you open a PR for the patch changes as well, and you can branch off this PR if you find any useful utility that can be used and that way, we add support incrementally for shadow DOM.

@yihuiliao and @snowystinger, Your feedback would be greatly appreciated to move things forward.

MahmoudElsayad commented:
@davidferguson One issue related to overlays, which I don't know if you have encountered yet or not, and that arose from accessibility testing, is that the `ariaHideOutside` function sets `aria-hidden="true"` on the document body, which is incorrectly inherited by elements within the shadow root making all overlays inherit `aria-hidden="true"`.

It's been handled in #6133; I am working on it getting the PR ready for review as well.

ritz078 commented:
@snowystinger @yihuiliao Can you have a look at this PR and suggest if there is something more that needs to be done ?

snowystinger commented:
We've had some other priorities. How ready is this PR for a review? I'll try to find some time this week to look at it again.

snowystinger commented:
It looks like there are some tests failing, do you need assistance figuring them out?

MahmoudElsayad commented:
@snowystinger I fixed the failing tests in the last commit.

MahmoudElsayad commented:
@snowystinger Sorry, the last commit doesn't handle versions 16 and 17, so I am adjusting the fix slightly; I am working on it now.

MahmoudElsayad commented:
@snowystinger They are fixed now.

snowystinger commented:
GET_BUILD

snowystinger commented on packages/@react-aria/interactions/test/useInteractOutside.test.js:
```diff
@@ -442,3 +442,150 @@ describe('useInteractOutside (iframes)', function () {
+    );
+
+    // Wait for dynamic element to be added
+    setTimeout(() => {
```
useFakeTimers, then run them just after this setTimeout, that way there's no danger of the assertions running after the test has completed

MahmoudElsayad replied:
Updated in [93a5071](https://github.com/adobe/react-spectrum/pull/6046/commits/93a5071440cc5e8e0dec313e287a816e525f1a8a).

MahmoudElsayad replied:
Updated in [93a5071](https://github.com/adobe/react-spectrum/pull/6046/commits/93a5071440cc5e8e0dec313e287a816e525f1a8a)

snowystinger commented on packages/@react-aria/interactions/test/useFocus.test.js:
```diff
@@ -153,4 +154,157 @@ describe('useFocus', function () {
+
+  describe('useFocus with Shadow DOM', function () {
+    it('handles focus events', function () {
+      const {shadowRoot, shadowHost} = createShadowRoot();
+      const events = [];
+      const ExampleComponent = () => (
+        <Example
+          onFocus={(e) => events.push({type: 'focus', target: e.target})}
+          onBlur={(e) => events.push({type: 'blur', target: e.target})}
+          onFocusChange={isFocused => events.push({type: 'focuschange', isFocused})} />
+      );
+
+      let root;
+      act(() => {
+        root = reactDomRenderer(<ExampleComponent />, shadowRoot);
+      });
```
I think you could restructure these and drop the reactDomRenderer if you did something along these lines

```
      const {shadowRoot, shadowHost} = createShadowRoot();
      const ExampleComponent = () => ReactDOM.createPortal(
        <Example
          onFocus={(e) => events.push({type: 'focus', target: e.target})}
          onBlur={(e) => events.push({type: 'blur', target: e.target})}
          onFocusChange={isFocused => events.push({type: 'focuschange', isFocused})} />,
         shadowRoot
      );
      let root = render(<ExampleComponent />);
```

Might also allow you to get rid of the extra act and it gives you a cleanup function on the react tree. It might also solve the setTimeout(..., 0) you have in here too

Unless the distinction of having the react root being inside the shadow dom matters?

MahmoudElsayad replied:
I updated the tests that used `reactDomRenderer` and removed it.

snowystinger reviewed (commented):
Got through tests today, I'll look more at the logic tomorrow hopefully.

@davidferguson 
> is there a preferred approach of how you'd like to see webcomponent & React Spectrum support progress

What would be best would be to understand the use case we're aiming to support. If you could provide information about your architecture and how you actually make use of shadowDOM, that'd be helpful. That way we can keep an eye out for potential pitfalls. Maybe an example app on github or in a codesandbox? (be mindful of what you make public though, you can reach out internally if it's sensitive)

Otherwise, a holistic approach is always good, but the work can be done in smaller chunks for ease of reviewing.

snowystinger reviewed (commented):
I noticed that FocusManager cannot move focus between shadow roots
```
it('should move focus forward across ShadowDOMs', async function () {
    let shadowRoot, unmount;
    let shadowRoot2, unmount2;
    function Test() {
      let ref = useRef(null);
      let [mounted, setMounted] = useState(false);
      let refCallback = useCallback((val) => {
        ref.current = val;
        setMounted(true);
        if (shadowRoot) {
          unmount();
        }
        if (shadowRoot2) {
          unmount2();
        }
        if (val) {
          let rv = createShadowRoot(val);
          let rv2 = createShadowRoot(val);
          shadowRoot = rv.shadowRoot;
          shadowRoot2 = rv2.shadowRoot;
          unmount = rv.unmount;
          unmount2 = rv2.unmount;
        }
      }, []);
      return (
        <div ref={refCallback}>
          <FocusScope>
            <Item data-testid="item1" />
            {mounted && ReactDOM.createPortal(<Item data-testid="item2" />, shadowRoot)}
            {mounted && ReactDOM.createPortal(<Item data-testid="item3" />, shadowRoot2)}
          </FocusScope>
        </div>
      );
    }

    function Item(props) {
      let focusManager = useFocusManager();
      let onClick = () => {
        focusManager.focusNext();
      };
      // eslint-disable-next-line jsx-a11y/click-events-have-key-events
      return <div {...props} tabIndex={-1} role="button" onClick={onClick} />;
    }

    let {getByTestId} = render(<Test />);
    let item1 = getByTestId('item1');
    let item2 = shadowRoot.querySelector('[data-testid="item2"]');
    let item3 = shadowRoot2.querySelector('[data-testid="item3"]');

    act(() => {item1.focus();});

    await user.click(item1);
    expect(getDeepActiveElement()).toBe(item2);

    await user.click(item2);
    expect(getDeepActiveElement()).toBe(item3);

    await user.click(item3);
    expect(getDeepActiveElement()).toBe(item3);
  });
```

I assume it's because of https://github.com/adobe/react-spectrum/blob/d7fc93273ec4956fbf3a48f18aee314e480de572/packages/%40react-aria/focus/src/FocusScope.tsx#L203 and that the tree walker won't go *into* a shadowroot, so you'll have to recurse into them yourself. It may be easier to build it into `getFocusableTreeWalker` and return an interface that matches the tree walker.

Likewise, restoring across shadowroots is going to be problematic with any checks for `contains`.

See comments from https://github.com/adobe/react-spectrum/issues/1472#issuecomment-761017517 for more things to keep an eye out for

I don't think we said there'd only be one shadowroot at the root of the entire application, so I'm assuming sibling/disjointed and nested are both valid.

MahmoudElsayad commented on packages/@react-aria/focus/test/focusSafely.test.js:
```diff
@@ -15,8 +15,16 @@ import {act, render} from '@react-spectrum/test-utils-internal';
+import {reactDomRenderer, unmount} from '@react-spectrum/test-utils-internal/src/reactCompat';
 import {setInteractionModality} from '@react-aria/interactions';
 
+function createShadowRoot() {
```
Updated in [93a5071](https://github.com/adobe/react-spectrum/pull/6046/commits/93a5071440cc5e8e0dec313e287a816e525f1a8a).

MahmoudElsayad commented on packages/@react-aria/utils/src/domHelpers.ts:
```diff
@@ -9,6 +9,50 @@ export const getOwnerWindow = (
+  return doc instanceof ShadowRoot ?  doc.ownerDocument.defaultView || window :  doc.defaultView || window;
+};
+
+export const getRootNode = (el: Element | null | undefined): Document | ShadowRoot => {
```
I think it makes sense to return null if the element is disconnected; I updated it in [93a5071](https://github.com/adobe/react-spectrum/pull/6046/commits/93a5071440cc5e8e0dec313e287a816e525f1a8a).

MahmoudElsayad commented on packages/@react-aria/interactions/test/usePress.test.js:
```diff
@@ -3301,4 +3302,640 @@ describe('usePress', function () {
+    function setupShadowDOMTest(extraProps = {}) {
+      const {shadowRoot, shadowHost} = createShadowRoot();
+      cleanupShadowRoot = shadowRoot;
+      cleanupShadowHost = shadowHost;
```
Updated in [93a5071](https://github.com/adobe/react-spectrum/pull/6046/commits/93a5071440cc5e8e0dec313e287a816e525f1a8a).

MahmoudElsayad commented on packages/@react-aria/interactions/test/usePress.test.js:
```diff
@@ -3301,4 +3302,640 @@ describe('usePress', function () {
+      const el = shadowRoot.getElementById('testElement');
+      fireEvent(el, pointerEvent('pointerdown', {pointerId: 1, pointerType: 'mouse'}));
+      fireEvent(el, pointerEvent('pointerup', {pointerId: 1, pointerType: 'mouse', clientX: 0, clientY: 0}));
+      expect(document.activeElement).not.toBe(el);
```
Used `getDeepActiveElement` in [93a5071](https://github.com/adobe/react-spectrum/pull/6046/commits/93a5071440cc5e8e0dec313e287a816e525f1a8a).

MahmoudElsayad commented on packages/@react-aria/interactions/test/usePress.test.js:
```diff
@@ -3301,4 +3302,640 @@ describe('usePress', function () {
+      expect(allowDefault).toBe(false);
+    });
+
+    it('should still prevent default when pressing on a non draggable + pressable item in a draggable container', function () {
```
Thanks for catching that, I should have added a wrapping div with `draggable` set to true in that test as I updated it [here](https://github.com/adobe/react-spectrum/pull/6046/commits/93a5071440cc5e8e0dec313e287a816e525f1a8a#diff-ce43d7ab73e0048f15b22b7575419a99290955b1c8cb39db0a00ddc77169455dR3739).

MahmoudElsayad commented on packages/@react-aria/focus/src/FocusScope.tsx:
```diff
@@ -133,7 +133,8 @@ export function FocusScope(props: FocusScopeProps) {
   // It cannot be a useLayoutEffect because the parent of this node hasn't been attached in the tree yet.
   useEffect(() => {
-    const activeElement = getOwnerDocument(scopeRef.current ? scopeRef.current[0] : undefined).activeElement;
+    // eslint-disable-next-line no-undef
```
Sorry, it was a leftover.

MahmoudElsayad commented on packages/@react-aria/utils/src/domHelpers.ts:
```diff
@@ -9,6 +9,50 @@ export const getOwnerWindow = (
 
-  const doc = getOwnerDocument(el as Element | null | undefined);
-  return doc.defaultView || window;
+  const doc = getRootNode(el as Element | null | undefined);
```
I am not sure I entirely follow; the Window case should be handled before the code reaches `getRootNode`.
```js
  if (el && 'window' in el && el.window === el) {
    return el;
  }
```

snowystinger replied:
Hmmm i don't recall what I was thinking for this, I think you're right. I might've just mixed them up in my head during my last review

snowystinger commented on packages/@react-aria/interactions/test/useInteractOutside.test.js:
```diff
@@ -572,19 +573,19 @@ describe('useInteractOutside shadow DOM extended tests', function () {
-      const dynamicEl = document.getElementById('dynamic-outside');
-      fireEvent.mouseDown(dynamicEl);
-      fireEvent.mouseUp(dynamicEl);
+    act(() => {jest.runAllTimers();});
```
do we actually need this? i think it's taken care of in the render's act now?

MahmoudElsayad commented on packages/@react-aria/focus/test/FocusScope.test.js:
```diff
@@ -1616,6 +1616,205 @@ describe('FocusScope', function () {
+
+  /**
+   * Test case: https://github.com/adobe/react-spectrum/issues/1472
+   * sandbox example: https://codesandbox.io/p/sandbox/vigilant-hofstadter-3wf4i?file=%2Fsrc%2Findex.js%3A28%2C30
```
For some reason, the sandbox is no longer available, I am removing the example.

MahmoudElsayad commented:
@snowystinger I extended the tree walker to handle traversal between multiple shadow DOMs by dynamically detecting and navigating to focusable elements in adjacent shadow roots. Please feel free to recheck the PR changes.

snowystinger commented:
Thanks, just a heads up, but the team is going on break. We'll be back in about a week. Until then it's unlikely we will look at this.

ritz078 commented:
@snowystinger Any update on this ?

snowystinger commented:
Thanks for the ping, I've not forgotten, I just haven't had a big enough solid chunk of time to dive back into this yet.

snowystinger commented on packages/@react-aria/focus/src/focusSafely.ts:
```diff
@@ -24,12 +24,12 @@ export function focusSafely(element: FocusableElement) {
+  const ownerDocument = getRootNode(element);
   if (getInteractionModality() === 'virtual') {
-    let lastFocusedElement = ownerDocument.activeElement;
+    let lastFocusedElement = ownerDocument?.activeElement;
```
why aren't we using getDeepActiveElement here and just below?

MahmoudElsayad replied:
You are right about it; using `getDeepActiveElement` is better, especially when it is related to nested shadow roots.

snowystinger commented on packages/@react-aria/interactions/src/useFocusVisible.ts:
```diff
@@ -135,23 +135,23 @@ function setupGlobalFocusEvents(element?: HTMLElement | null) {
-  documentObject.addEventListener('keydown', handleKeyboardEvent, true);
-  documentObject.addEventListener('keyup', handleKeyboardEvent, true);
-  documentObject.addEventListener('click', handleClickEvent, true);
+  documentObject?.addEventListener('keydown', handleKeyboardEvent as EventListener, true);
```
why did the cast become necessary?

MahmoudElsayad replied:
It is necessary because there is a gap between specific event types (e.g., `KeyboardEvent`) and the `EventListener` type expected by `addEventListener`. I updated the approach to use [createEventListener](https://github.com/adobe/react-spectrum/pull/6046/commits/40cd5dd2e6f88d02680fb847a2500d4bbecff51a#diff-c85095437b8c2976c74228b10139d930515efcd0be5bef41bda3160ba91ac2dbR38-R41) that maintains type safety and existing function signatures without casting.

MahmoudElsayad replied:
I believe that the approach I mentioned above will cause problems with cleanup, so I updated it [here](https://github.com/adobe/react-spectrum/pull/6046/commits/c5ddb2423ac07abbcd036391218fc23bb65acfee).

snowystinger commented on packages/@react-aria/interactions/src/useFocusWithin.ts:
```diff
@@ -70,7 +71,7 @@ export function useFocusWithin(props: FocusWithinProps): FocusWithinResult {
     // Double check that document.activeElement actually matches e.target in case a previously chained
     // focus handler already moved focus somewhere else.
-    if (!state.current.isFocusWithin && document.activeElement === e.target) {
+    if (!state.current.isFocusWithin && getDeepActiveElement() === e.target) {
```
down below you say that e.target will be the shadowRoot, but won't getDeepActiveElement find the element inside the shadowRoot?

snowystinger commented on packages/@react-aria/interactions/src/usePress.ts:
```diff
@@ -303,7 +317,9 @@ export function usePress(props: PressHookProps): PressResult {
             };
 
-            addGlobalListener(getOwnerDocument(e.currentTarget), 'keyup', chain(pressUp, onKeyUp), true);
+            const ownerDocument = getRootNode(e.currentTarget) || getOwnerDocument(e.currentTarget);
```
isn't this equivalent?
```
const ownerDocument = getRootNode(e.currentTarget);
```
or when does it fallback to getOwnerDocument?

MahmoudElsayad replied:
I removed the fallback, it will never fallback as long as there is a `e.currentTarget`.

MahmoudElsayad commented on packages/@react-aria/utils/src/domHelpers.ts:
```diff
@@ -9,6 +9,49 @@ export const getOwnerWindow = (
+  }
+
+  // Return the root node, which can be either a Document or a ShadowRoot
+  return rootNode instanceof Document ? rootNode : rootNode as ShadowRoot;
```
Typescript will complain as `getRootNode` returns `Node` that will change based on where it is called. I updated the snippet to use type guards instead of casting.

MahmoudElsayad commented:
@snowystinger The failing tests didn't work as soon as I got a pull from the latest master before committing any updates to the PR. Is that expected?

snowystinger commented:
In general, no, but I took a look, and the failures appear to all be testing the stuff you're working on. So I'd double-check your merge or that something on main didn't conflict with the changes you're trying to make. We are working on a lot of things, I cannot guarantee everything we do on main will be compatible with what you're doing. This PR is very ambitious.

I'll have a look at your latest changes soon.

MahmoudElsayad commented:
Thanks for taking the time to look into this. I appreciate your efforts in reviewing it thoroughly 🙏🏼. I understand it is a substantial change and can conflict with many things. I'll look carefully at potential conflicts with recent updates to the main branch; earlier, I wanted to confirm that it was unexpected.

MahmoudElsayad commented:
@snowystinger The tests failed because the previous implementation used `instanceof` before your recommendation for changes [here](https://github.com/adobe/react-spectrum/pull/6046#discussion_r1678758430). `instanceof` checks fail across contexts due to different global constructors, by switching to `nodeType` checks, which are consistent across contexts, we ensure that the function accurately identifies Document and ShadowRoot nodes regardless of their origin.

ritz078 commented:
@snowystinger Any chance you can look into this ? It's getting hard for us to upgrade with so many patches in our projects. Also, we want to open more PRs on top of this.

snowystinger commented on packages/@react-aria/interactions/src/useFocusVisible.ts:
```diff
@@ -210,15 +228,17 @@ const tearDownWindowFocusTracking = (element, loadListener?: () => void) => {
-  if (documentObject.readyState !== 'loading') {
+
+   // Shadow root doesn't have a readyState, so we can assume it's ready in case of there is a shadow root.
+  if (rootNode instanceof ShadowRoot || (rootNode?.readyState !== 'loading')) {
```
will this have a problem with `instanceof`?

MahmoudElsayad replied:
Yes, it can be problematic, and it is better to update it with `nodeType`; I shall do the update.

MahmoudElsayad replied:
Update in the last commit.

snowystinger commented on packages/@react-aria/focus/src/FocusScope.tsx:
```diff
@@ -742,10 +745,12 @@ function restoreFocusToElement(node: FocusableElement) {
   let selector = opts?.tabbable ? TABBABLE_ELEMENT_SELECTOR : FOCUSABLE_ELEMENT_SELECTOR;
-  let walker = getOwnerDocument(root).createTreeWalker(
-    root,
+  let doc = root instanceof ShadowRoot ? root :  (getRootNode(root) || getOwnerDocument(root));
```
will all of these `instanceof`'s have issues?

MahmoudElsayad replied:
Updated in the last commit.

snowystinger reviewed (dismissed):
This looks pretty good, I'm giving a provisional approval (couple small last questions). Will get someone else to review it now.

snowystinger commented:
looks like the accidental rebase on main has affected this, you may need to run `git rebase origin/main`
@devongovett ^

alexandersorokin commented:
Hello everyone!

This MR solves almost all issues in my setup, but it introduces two new issues.
My setup is executes rendering code in an same-origin iframe and renders DOM in another same-origin iframe. React itself and many other libraries support this scenario. For example, it is described in [Shadow realm proposal](https://github.com/tc39/proposal-shadowrealm/blob/main/explainer.md#Iframes).

The two remaining issues are the following:
1. Using `instanceof`. Different iframes have different constructors. Could we change `instanceof Document` to `instanceof el.ownerDocument.defaultView.document`? Or get rid of `instanceof` at all?
2. Using `document.activeElement` in `getDeepActiveElement` function. Could we get an active element from any element by using `someElement.ownerDocument.activeElement` method?

It would be good using as few global objects as possible.

MahmoudElsayad commented:
@alexandersorokin I updated most of the usage of `instanceof` to depend on `nodeType` instead, and it fixed all tests available for iframes. If you know of a specific a place where the use of `instanceof` is causing an issue, can you try updating the code to use `nodeType` and see if it will fix the issues?

alexandersorokin commented:
@MahmoudElsayad, thank you for your reply and your efforts of pushing forward with MR.

There are good examples of how to use `instanceof` in the repo: https://github.com/adobe/react-spectrum/blob/e1b72a79f9cc131d8ac50bb5d89ef1f42a4e01b4/packages/%40react-aria/focus/src/isElementVisible.ts#L17
It uses [getOwnerWindow](https://github.com/adobe/react-spectrum/blob/e1b72a79f9cc131d8ac50bb5d89ef1f42a4e01b4/packages/%40react-aria/utils/src/domHelpers.ts#L5). Using `nodeType` is an even better solution (it supports more scenarios). But both work in my case.

In my previous message, I would like to ask you not to introduce new usages of `instanceof` without using `getOwnerWindow` because they will not work in same-origin iframe scenarios. Currently, there are several `doc instanceof ShadowRoot` in this MR are introduced. Could you replace them with `doc.nodeType === Node.DOCUMENT_FRAGMENT_NODE && 'host' in doc` or with `doc instanceof getOwnerWindow(doc).ShadowRoot`?

There is no necessity to fix existing `instanceof` occurrences.

alexandersorokin commented on packages/@react-aria/focus/src/focusSafely.ts:
```diff
@@ -24,12 +24,12 @@ export function focusSafely(element: FocusableElement) {
   // causing the page to scroll when moving focus if the element is transitioning
   // from off the screen.
-  const ownerDocument = getOwnerDocument(element);
+  const activeElement = getDeepActiveElement();
```
In order to continue to work in same-origin iframe scenario I suggest to get `document`  instance in ` getDeepActiveElement` from `element.ownerDocument` (not to use global `document` instance).

alexandersorokin replied:
resolved

MahmoudElsayad commented:
@snowystinger I've updated the usages of `instanceof` to depend on `nodeType`.

@alexandersorokin ^

snowystinger commented on packages/@react-aria/interactions/src/useFocusVisible.ts:
```diff
@@ -210,15 +228,24 @@ const tearDownWindowFocusTracking = (element, loadListener?: () => void) => {
+    // If it's a Document that's ready or a ShadowRoot
     setupGlobalFocusEvents(element);
-  } else {
+  } else if (rootNode?.nodeType === Node.DOCUMENT_NODE) {
```
```suggestion
  } else if (isDocument) {
```

snowystinger commented on packages/@react-aria/focus/src/FocusScope.tsx:
```diff
@@ -771,6 +789,13 @@ export function getFocusableTreeWalker(root: Element, opts?: FocusManagerOptions
     walker.currentNode = opts.from;
   }
 
+  if (doc instanceof ShadowRoot) {
```
missed instanceof

snowystinger commented on packages/@react-aria/focus/src/FocusScope.tsx:
```diff
@@ -873,6 +898,46 @@ function last(walker: TreeWalker) {
+    let nextElement = originalNextNode();
+    if (!nextElement && scope && scope.length > 0) {
+      let currentShadowRoot = scope[0].getRootNode();
+      let nextShadowHost = currentShadowRoot instanceof ShadowRoot ? currentShadowRoot.host.nextElementSibling : null;
```
missed instanceof

snowystinger commented on packages/@react-aria/focus/src/FocusScope.tsx:
```diff
@@ -873,6 +898,46 @@ function last(walker: TreeWalker) {
+    let previousElement = originalPreviousNode();
+    if (!previousElement && scope && scope.length > 0) {
+      let currentShadowRoot = scope[0].getRootNode();
+      let previousShadowHost = currentShadowRoot instanceof ShadowRoot ? currentShadowRoot.host.previousElementSibling : null;
```
same

snowystinger commented on packages/@react-aria/interactions/src/useFocusVisible.ts:
```diff
@@ -210,15 +228,24 @@ const tearDownWindowFocusTracking = (element, loadListener?: () => void) => {
   let loadListener;
-  if (documentObject.readyState !== 'loading') {
+
+  const isDocument = rootNode?.nodeType === Node.DOCUMENT_NODE;
+  const isShadowRoot = rootNode?.nodeType === Node.DOCUMENT_FRAGMENT_NODE && 'host' in rootNode;
```
Looks like these should be off in the helper functions

snowystinger reviewed (commented):
thanks for the update and again, sorry about the merge issues
I should have another teammate to review this soon

MahmoudElsayad commented:
@snowystinger I replaced all usages of `instanceof` now in the PR and created some utils for it.

alexandersorokin commented:
> @snowystinger I've updated the usages of `instanceof` to depend on `nodeType`.
> 
> @alexandersorokin ^

The first issue about `instanceof` is resolved.

The second issue is not resolved because the `getDeepActiveElement` function doesn't receive document from `ownerDocument`.

MahmoudElsayad commented:
> > @snowystinger I've updated the usages of `instanceof` to depend on `nodeType`.
> > 
> > @alexandersorokin ^
> 
> The first issue about `instanceof` is resolved. Thank you.
> 
> The second issue is not resolved because the `getDeepActiveElement` function doesn't receive document from `ownerDocument`.

@alexandersorokin Can you write a test for the second issue and I will check the best approach to handle it, as I will have to do some refactoring for the `getDeepActiveElement` that way.

alexandersorokin commented:
> > > @snowystinger I've updated the usages of `instanceof` to depend on `nodeType`.
> > > @alexandersorokin ^
> > 
> > 
> > The first issue about `instanceof` is resolved. Thank you.
> > The second issue is not resolved because the `getDeepActiveElement` function doesn't receive document from `ownerDocument`.
> 
> @alexandersorokin Can you write a test for the second issue and I will check the best approach to handle it, as I will have to do some refactoring for the `getDeepActiveElement` that way.

```
const iframe = document.createElement("iframe");
iframe.src = "about:blank";
await new Promise((resolve) => {
  iframe.onload = resolve;
  document.body.appendChild(iframe);
});


const childDocument = iframe.contentWindow.document;

const input1 = childDocument.createElement('input');
const input2 = childDocument.createElement('input');
const input3 = childDocument.createElement('input');

childDocument.body.appendChild(input1);
childDocument.body.appendChild(input2);
childDocument.body.appendChild(input3);

input2.focus()

expect(getDeepActiveElement()).toBe(input2);
```

Currently `getDeepActiveElement()` returns `iframe` instead of `input2`.

The example is artificial. In real life scripts are executed inside iframe but DOM is rendered outside iframe.

-------------------------

To work correctly `getDeepActiveElement` may accept `document` as argument:
```
const getDeepActiveElement = (doc: Document) => {
  let activeElement = doc.activeElement;
  while (activeElement?.shadowRoot && activeElement.shadowRoot?.activeElement) {
    activeElement = activeElement.shadowRoot.activeElement;
  }
  return activeElement;
};
```

Caller of `getDeepActiveElement` can get document from `ownerDocument`.

MahmoudElsayad commented:
@alexandersorokin I updated the `getDeepActiveElement` function as it was needed as well to fix a bug where whenever you open an overlay using a trigger and close the overlay, the focus doesn't return to the trigger in shadow DOM. It is now fixed in this [commit](https://github.com/adobe/react-spectrum/pull/6046/commits/54cfe242704b59e01003cc7ef83717ae9d5d8899).

snowystinger reviewed (dismissed):
Did you want to include the test case outlined here?
https://github.com/adobe/react-spectrum/pull/6046#issuecomment-2438880004
or how are we making sure we don't regress on that?

alexandersorokin commented:
> @alexandersorokin I updated the `getDeepActiveElement` function as it was needed as well to fix a bug where whenever you open an overlay using a trigger and close the overlay, the focus doesn't return to the trigger in shadow DOM. It is now fixed in this [commit](https://github.com/adobe/react-spectrum/pull/6046/commits/54cfe242704b59e01003cc7ef83717ae9d5d8899).

I appreciate your efforts in order to make the good MR. The `getDeepActiveElement` function now accepts `document`. Thank you. That is the best way of restoring the functionality of working with iframes.

But currently only one of four calls of `getDeepActiveElement` uses `ownerDocument`. It is: https://github.com/adobe/react-spectrum/blob/e918e38d4e9e77025c70ae5a81f75ddbc9f60fb5/packages/%40react-aria/focus/src/FocusScope.tsx#L571 .

Other three of four calls of `getDeepActiveElement` still don't use `ownerDocument`:
1. https://github.com/adobe/react-spectrum/blob/e918e38d4e9e77025c70ae5a81f75ddbc9f60fb5/packages/%40react-aria/focus/src/focusSafely.ts#L27
2. https://github.com/adobe/react-spectrum/blob/e918e38d4e9e77025c70ae5a81f75ddbc9f60fb5/packages/%40react-aria/interactions/src/useFocus.ts#L67
3. https://github.com/adobe/react-spectrum/blob/e918e38d4e9e77025c70ae5a81f75ddbc9f60fb5/packages/%40react-aria/interactions/src/useFocusWithin.ts#L74

But two of three use `ownerDocument` in the `main` branch:
1. https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/focus/src/focusSafely.ts#L27
2. https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/interactions/src/useFocus.ts#L66

In order to eliminate regression, could you update them in the way so `ownerDocument` is passed to `getDeepActiveElement` here? So it will match `main` branch.

The test I have sent before doesn't check all cases. It shows a general way of working with same-origin iframes. In order to work correctly, all instances of `getDeepActiveElement` that used `ownerDocument` before MR should use `ownerDocument` after MR.

MahmoudElsayad commented:
@snowystinger I added a test for the use case mentioned and changed the usages of `getDeepActiveElement` to depend on the document, as @alexandersorokin also discussed.

snowystinger commented:
Nothing else right now, I'm just trying to get you a second reviewer.

LFDanLu commented:
@snowystinger has pushed up some fixes from the team's initial internal review where we noticed that `usePress` wouldn't clean up when pointer down started in the shadowDOM but pointer up happens outside it. Additionally we found that `e.target` returns the shadowRoot itself instead of the actual element within and fixed that. Finally, `node.contains` doesn't work across shadowDOM boundaries so we've updated that as well.

We'll continue reviewing the other parts of this PR for similar concerns, let us know if you have any questions.

dielduarte commented:
Hey, team! First, thanks for all your work on this and the entire ecosystem! My team @appcues has been testing a couple of options, and react aria fits perfectly for what we need, though we hit into this issue as our extension is wrapped within a shadow-dom element. 

I'm wondering if my team can do anything to help with this work, pls lmk! I don't want to pressure you all, but betting on react-aria seems like the best option for our future projects.

Thank you!

snowystinger commented:
Fixed esm locally, looking into react 19 failure

snowystinger commented:
@MahmoudElsayad @ritz078 would you mind testing the new changes in your application and let us know if they still work for you

LFDanLu commented on packages/@react-aria/focus/src/FocusScope.tsx:
```diff
@@ -508,7 +515,7 @@ function useAutoFocus(scopeRef: RefObject<Element[] | null>, autoFocus?: boolean
       activeScope = scopeRef;
       const ownerDocument = getOwnerDocument(scopeRef.current ? scopeRef.current[0] : undefined);
-      if (!isElementInScope(ownerDocument.activeElement, activeScope.current) && scopeRef.current) {
+      if (!isElementInScope(getActiveElement(ownerDocument), activeScope.current) && scopeRef.current) {
```
@snowystinger I'm still seeing focus get lost to the body when testing "Focusable First in Scope" story with Android Talkback. Digging to see what exactly is happening but doesn't seem to get to this part of the code

devongovett commented on packages/@react-aria/interactions/src/usePress.ts:
```diff
@@ -279,8 +293,8 @@ export function usePress(props: PressHookProps): PressResult {
       onKeyDown(e) {
-        if (isValidKeyboardEvent(e.nativeEvent, e.currentTarget) && e.currentTarget.contains(e.target as Element)) {
-          if (shouldPreventDefaultKeyboard(e.target as Element, e.key)) {
+        if (isValidKeyboardEvent(e.nativeEvent, e.currentTarget) && nodeContains(e.currentTarget, e.nativeEvent.composedPath()[0] as Element)) {
```
Can we make a util for this, e.g. `getEventTarget`? Not sure but this may also be slower than accessing e.target since the browser will have to construct an array...

devongovett commented on packages/@react-aria/utils/src/domHelpers.ts:
```diff
@@ -12,3 +12,83 @@ export const getOwnerWindow = (
+ * Type guard that checks if a node is a Document node. Uses nodeType for cross-context compatibility.
+ */
+export function isDocument(node: Node | null): node is Document {
+  return isNode(node) && node.nodeType === Node.DOCUMENT_NODE;
```
Is the `isNode` check really necessary? You already know it's a node from the type declaration. TypeScript will ensure that.

devongovett commented on packages/@react-aria/utils/src/shadowdom/DOMFunctions.ts:
```diff
@@ -0,0 +1,85 @@
+// Source: https://github.com/microsoft/tabster/blob/a89fc5d7e332d48f68d03b1ca6e344489d1c3898/src/Shadowdomize/DOMFunctions.ts#L16
```
License needs to be added to NOTICE file

devongovett commented on packages/@react-aria/utils/src/shadowdom/DOMFunctions.ts:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    if (
+          typeof (currentNode as HTMLSlotElement).assignedElements !==
```
can we format this a bit nicer?

devongovett replied:
also maybe better to check `currentNode.tagName !== 'SLOT'` instead of looking for this method?

devongovett commented on packages/@react-aria/utils/src/shadowdom/DOMFunctions.ts:
```diff
@@ -0,0 +1,85 @@
+      ) {
+          // Element is slotted
+      currentNode = (currentNode as HTMLElement).assignedSlot?.parentNode;
+    } else if (currentNode.nodeType === document.DOCUMENT_FRAGMENT_NODE) {
```
Use the `isShadowRoot` util. Then you don't need the TS cast below.

devongovett commented on packages/@react-aria/utils/src/shadowdom/DOMFunctions.ts:
```diff
@@ -0,0 +1,85 @@
+    if (
+          typeof (currentNode as HTMLSlotElement).assignedElements !==
+              'function' &&
+          (currentNode as HTMLElement).assignedSlot?.parentNode
```
Looks like `assignedSlot` is a property of `Element` not `HTMLElement`

devongovett commented on packages/@react-aria/utils/src/shadowdom/ShadowTreeWalker.ts:
```diff
@@ -0,0 +1,268 @@
+    this._currentNode = root;
+
+    this._walkerStack.unshift(
+            doc.createTreeWalker(root, whatToShow, this._acceptNode)
```
formatting?

devongovett commented on packages/@react-aria/utils/src/shadowdom/ShadowTreeWalker.ts:
```diff
@@ -0,0 +1,268 @@
+    return null;
+  }
+}
+
```
Should we have unit tests for this?

snowystinger replied:
good call, I thought they had tests, but I couldn't find any, so I've written at least a few to get us started.

devongovett commented on packages/@react-spectrum/list/test/ListView.test.js:
```diff
@@ -980,27 +980,26 @@ describe('ListView', function () {
         expect(onAction).not.toHaveBeenCalled();
       });
```
why were there test changes necessary?

snowystinger replied:
We had a lot of tests that were not true to real life. Firing keyboard events without focus being on the element for example. The ShadowTreeWalker enforces a certain condition in order to work, which is totally fine in the real world, but we could and did easily break in testing. 
https://github.com/adobe/react-spectrum/pull/6046/files#diff-582a0300fa3cc6033754f65cd6965774fe194ab0b85211e4e400eef0656ecc46R78

LFDanLu replied:
we'll probably want to call that out when we release this so that others who've done the same in their own tests know that they need to update their tests in a similar fashion since it is non-obvious why their tests may suddenly start failing otherwise

devongovett reviewed (commented):
I have two primary concerns:

1. The bundle size impact of adding all these extra utils, even for the 99% of applications that never use shadow dom. Can we measure what this comes out to? Should we externalize these somehow so they can be loaded only when needed?
2. Performance. Using native DOM methods like `contains` and `TreeWalker` will be faster than manually implemented versions. Would be nice to measure this as well somehow. I'd look at things like drag and drop on a large DOM, e.g. in the docs.

Ideally supporting shadow dom would only impact teams that actually use it, and not everyone else. Open to ideas on this front.

snowystinger commented:
> Ideally supporting shadow dom would only impact teams that actually use it, and not everyone else. Open to ideas on this front.

One idea was
```
let DomContext = createContext({
  nodeContains: (x, y) => x.contains(y),
  treeWalker: (node) => document.createTreeWalker(),
  ...
})

// App using shadows
import {ShadowDOM} from '@react-aria/...';
let App = () => (
  <DomContext.Provider value={ShadowDOM}>...</...>
)
```
Where import ShadowDOM would be all the equivalent functions that have Shadow support

It would be tough in some utilities which are not hooks or components, such as ariaHideOutside... but we could add an extra optional options argument to the function with the same defaults as the Context defaults

devongovett commented:
Yeah maybe it could even be a global registry rather than a context. Not sure if it would be necessary to scope to a particular subtree.

LFDanLu commented on packages/@react-spectrum/table/test/TableDnd.test.js:
```diff
@@ -3053,24 +3077,22 @@ describe('TableView', function () {
       draghandle = within(rows[0]).getAllByRole('button')[0];
       expect(draghandle).toBeTruthy();
       expect(draghandle).toHaveAttribute('draggable', 'true');
-      moveFocus('ArrowRight');
-      fireEvent.keyDown(draghandle, {key: 'Enter'});
-      fireEvent.keyUp(draghandle, {key: 'Enter'});
+      await user.keyboard('{ArrowRight}');
+      await user.keyboard('{Enter}');
```
Nit: the intent of this portion of the test was to make sure that triggering drag on a non-selected item didn't include selected rows (i.e. it selects the row with `key='8'` then start a drag via row with `key='0'` and ensures that drag start only includes `key='0'`). As it is now, this drags the same row that is selected which is already tested above. I'll make the comment above clearer and see about making this change after finishing my sweep through the other tests

LFDanLu commented on packages/@react-aria/interactions/src/useFocusWithin.ts:
```diff
@@ -17,6 +17,7 @@
 
 import {DOMAttributes} from '@react-types/shared';
 import {FocusEvent, useCallback, useRef} from 'react';
+import {getActiveElement, getOwnerDocument} from '@react-aria/utils';
```
Can be done later, but would be nice to have a lint rule or something that makes sure that we make use of these instead of using `document.activeElement`

LFDanLu commented on packages/@react-aria/utils/src/shadowdom/ShadowTreeWalker.ts:
I remember when we did some test debugging for this last week that there were some changes that needed to be made from Microsoft's original implementation, are those documented anywhere? Would be nice to know in case we need to compare again in the future

snowystinger replied:
No, but I included the git commit that I pulled the code from, so you could theoretically run a diff on the two.

As for why, I needed to implement firstChild/lastChild
Also, in the course of writing tests I found a couple bugs, I've now addressed those that I found. Mostly in next/prev node

LFDanLu reviewed (commented):
Did a sweep of the tests and the changes looks reasonable, will push up some minor changes to a couple. I noticed that there are still calls `document.activeElement` and `.contains` elsewhere in the codebase, I assume the plan is to eventually convert those over to use the domHelpers?

snowystinger commented:
perf and size comments on PR https://github.com/adobe/react-spectrum/pull/7548#issuecomment-2594347745

ritz078 commented:
> would you mind testing the new changes in your application and let us know if they still work for you

@snowystinger These changes look good so far. Our internal tests are passing. We will be opening some small followup PRs too once this gets merged. Any idea about the timeline for this PR ?

snowystinger commented:
Thanks for checking in, we are still in discussions internally. Should have more information for you early next week.

devongovett reviewed (approved):
Thanks for working on this. I appreciate the persistence, and I know it's been a long process.

In the future, I think we may need to revisit how some of our components are implemented more holistically to account for shadow dom. While deeply traversing into shadow roots works in some cases, it won't work with closed shadow roots. For example, even with the FocusScope changes here, things like native `<video>` elements, `<input type="date">`, etc., along with any third party components using closed shadow roots will still not work correctly since we cannot traverse into them. Fixing that would require a completely different implementation strategy.

Overall, shadow dom is meant to encapsulate behavior, and deep traversal sort of goes against that goal. For now it's better than nothing, but it would be good to try to find ways to avoid it in the future if possible.

snowystinger commented on NOTICE.txt:
```diff
@@ -211,3 +211,32 @@ This codebase contains a modified portion of code from Yarn berry which can be o
+
+
+-------------------------------------------------------------------------------
+This codebase contains a modified portion of code from Yarn berry which can be obtained at:
```
I think i copy pasted that
```suggestion
This codebase contains a modified portion of code from Microsoft which can be obtained at:
```

snowystinger commented:
Thanks again, this will be available in our nightlies starting tonight. You can enable the behavior using the feature flag.
```
import {enableShadowDOM} from '@react-stately/flags';
```

As @devongovett noted, we're starting to think about how we want to move forward with this kind of work. We've made a start here: https://github.com/adobe/react-spectrum/pull/7687
Feel free to add your voice as well. We're still building up the requirements and it helps to have real life cases to draw from.
